docs/manual/tsort.md

```diff
@@ -2,160 +2,56 @@
-Requires: from added packages.
-
-Except for the "well known" libtermcap <-> bash loop (which is just wrong),
-all of the other dependencies are simply not needed in an upgrade context
-to perform package ordering. Please note that all of the known to cause
-loop dependencies listed above are, for now, explicitly ignored when
-determining package install ordering.
-
-## Summary
-
-So what does this all mean? Basically not much, unless you find yourself
-trying to specify dependencies amongst a set of packages correctly and
-happen to create a dependency loop.
-
-And, before you start adding the new-fangled syntax to packages, please
-remember that rpm will almost certainly be auto-generating fine-grained
-dependencies for %post et al scriptlets pretty soon. Truly, rpm needs to
-make packaging easier, not provide Yet More Complicated Syntax in spec files.
-
-With thanks to Ken Estes for doing the implementation in bash2 that makes
-it possible to auto-generate scriptlet dependencies, blame me for the long,
-slow deployment.
+# Package ordering
+
+Rpm orders transactions solely based on package dependency information:
+dependencies first on install, last on erase. If the install/erase
+order is not right, then either the dependency information is incomplete,
+incorrect or there are unresolvable dependency loops.
+
+`Requires` are the primary data for ordering. If package A Requires B,
+then provider of B will be installed before A, and erased after it.
+Weak dependencies (`Recommends`, `Suggests, `Supplements, `Enhances`)
+and ordering hints (`OrderWithRequires`) are simiarly considered, if
+the provider is present in the transaction. Note that `Supplements` and
+`Enhances` are reverse dependencies and are ordered thus.
+
+A dependency loop is generally a packaging bug as no correct ordering is
+possible to achieve in that situation. They are best solved by splitting
+to sub-packages to eliminate the loop, but sometimes this is not practical,
+and also some dependencies are not relevant for ordering at all. Rpm
+supports various [dependency qualifiers](spec.md#Requires) to supply
+context which rpm can and does use to improve ordering and resolving
+dependency loops.
+
+Qualifiers affecting install ordering:
+* `pre`
+* `post`
+
+Qualifiers affecting erase ordering only:
+* `preun`
+* `postun`
+
+Normally, `Requires: A` and `Requires(pre): A` result in exactly the same
+order but when a loop is detected, a dependency with an ordering-qualifier
+are favored.
+
+Both install and erase ordering qualifiers are additionally used as loop
+cutting hints in presence of dependency loops. Packages belonging to a
+dependency loop are installed as a consequtive group, using the qualifier
+hints to try and cut the loop with least minimal disruption, but one should
+remember that there is no such thing as a correct order for a loop.
+
+Qualifiers denoting an unordered dependency, ie dismissing a dependency from
```

Comment:
Too complicated language! May be "Dependencies with the following qualifiers are ignored for ordering:"